src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp

```diff
@@ -156,12 +159,11 @@ class G1BlockOffsetTablePart {
 
   void verify() const;
 
-  // Returns the address of the start of the block containing "addr", or
-  // else "null" if it is covered by no block.  (May have side effects,
-  // namely updating of shared array entries that "point" too far
-  // backwards.  This can occur, for example, when lab allocation is used
-  // in a space covered by the table.)
-  inline HeapWord* block_start(const void* addr);
+  // Returns the address of the start of the block containing "addr".
+  inline HeapWord* block_start(const void* addr) const;
+  // Returns the address of the start of the block containing "addr", assuming that
+  // the given address is card-aligned.
+  inline HeapWord* block_start_aligned(const void* addr) const;
```

Comment:
I am not sure introducing another API is really that necessary. What I have in mind is sth like:

```c
void block_start(const void* addr) {
  if (card_aligned(addr)) {
    fast_path
  } else {
    slow_path
  }
  assert_post-cond
}
```

I don't think the cost of if-check will be observable.

Reply:
I had something like that, but the "not observable" part was not that clear in my results (i.e. during garbage collection) so I only optimized the part where this can be statically determined for now.

Reply:
CPU resources like branch prediction/speculative execution aren't free either, and this is a very performance sensitive area of code.

Reply:
Then why not just card-align addr and always use the fast path code?

Reply:
@kimbarrett: The `block_start` method returns the first object reaching into `addr`; if we aligned first, then the result would be the first object reaching into `aligned(addr)` which is different; the callers would need to advance forward not-requested objects then.
Or maybe I'm missing something in your suggestion?

Reply:
> I had something like that, but the "not observable" part was not that clear in my results

I added the following to `G1BlockOffsetTablePart::block_start` to see the cost.

```c
  if (is_aligned(addr, BOTConstants::card_size())) {
    return q;
  }
```

Using `Scan Heap Roots (ms)` and `Scanned Cards` from `-Xlog:gc,gc+phases`, I can calculate #cards processed per ms. Surprisingly, I can observe some reduction (~2.7%) in this metric on my box.

Re Kim's suggestion, the model where `G1BlockOffsetTablePart::block_start` accepts only card-aligned addresses and the caller (`HeapRegion::block_start`) handles block-walking (heap parsing) makes more sense to me. (After all, "block size" is sth meaningful only inside `HeapRegion`.) Ofc, this can be dealt with in this PR or another.

Reply:
[Here](https://github.com/openjdk/jdk/compare/master...tschatzl:jdk:submit/move-block-size-out-of-bot?expand=1)'s a prototype for (trying to) move `block_start` out of `G1BlockOffsetTablePart`; in the end I have a bit of mixed feelings about it:
* the BlockOffsetTable deals with blocks; blocks intrinsically have a start and a size; removing the BlockOffsetTable to only deal with parts of a block seems a bit limited.
Additionally not knowing about size will require to move out `verify` too if we are strict about this distinction, which means that lots of code is being dumped in the already (too) large `HeapRegion` class. Of course, like in the prototype, the `G1BlockOffsetTablePart::verify` could also stay as is....
* the BlockOffsetTable is already very tightly dependent on `HeapRegion` anyway (for current layout, i.e. `bottom`, `top`, `end`).

Comments?